test/4_test.py

```diff
@@ -0,0 +1,315 @@
+                },
+                headers = {
+                    'Authorization': self.access_token
+                }
+            ).json()
+            if response.get('event_id', None) is None:
+                print('Message was rejected immediately')
+            else:
+                # Message may be redacted later
+                bad_events[response['event_id']] = "Good image with bad description, type %s" % message_type
+
+        print('Sending canary event, to ensure that all previous events have been flushed')
+        response = requests.put('http://localhost:8080/_matrix/client/r0/rooms/%s/send/m.room.message/canary-%s' % (quoted_room_id, uuid.uuid1()),
+            json = {
+                'body': 'A text without any link',
+                'format': 'org.matrix.custom.html',
+                'formatted_body': 'A rich text without any link',
+                'msgtype': 'm.text'
+            },
+            headers = {
+                'Authorization': self.access_token
+            }
+        ).json()
+        canary_event = response['event_id']
+
+        # Wait until the canary has been treated.
+        since = None
+        while True:
+            response = requests.get('http://localhost:8080/_matrix/client/r0/sync', json = {
+                'since': since
+            }, headers = {
+                    'Authorization': self.access_token
+                }
+            ).json()
+            since = response.get('since', None)
+            found_canary = False
+            for id in self._item_generator(response, 'event_id'):
+                if id == canary_event:
+                    found_canary = True
+                    break
+            if found_canary:
+                break
+            # Otherwise, wait a little.
+            print("Apparently, not all events have been flushed, waiting a bit")
+            time.sleep(0.5)
+
+        # Now that we're sure that all events have been treated, re-download all events
+        # and make sure that the good ones are all present and the bad ones are all either
+        # absent or redacted away.
+        print("Now checking that all good events have passed and all bad events have been rejected")
+        since = None
+        bad_events_found = set()
+        while True:
+            response = requests.get('http://localhost:8080/_matrix/client/r0/sync', json = {
+                'since': since
+            }, headers = {
+                    'Authorization': self.access_token
+                }
+            ).json()
+            since = response.get('since', None)
+            found_canary = False
+
+            # Ensure that all the good events are present
+            for id in self._item_generator(response, 'event_id'):
+                print("Encountering event id %s" % id)
+                if id == canary_event:
+                    found_canary = True
+                maybe_good_event = good_events.get(id, None)
+                if maybe_good_event != None:
```

Comment:
This would be clearer if it was `if id in good_events` instead of fetching it checking it against `None`. (Same for the bad events below.)